zendesk/resource_zendesk_trigger_test.go

```diff
@@ -1,22 +1,201 @@
+		t.Fatalf("trigger had title value %v. should have been %v", trg.Title, v)
+	}
+}
+
+func TestCreateTrigger(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	m := mock.NewClient(ctrl)
+	i := newIdentifiableGetterSetter()
+	out := zendesk.Trigger{
+		ID:    12345,
+		Title: "trigger",
+	}
+
+	m.EXPECT().CreateTrigger(gomock.Any(), gomock.Any()).Return(out, nil)
+	if err := createTrigger(i, m); err != nil {
+		t.Fatal("CreateTrigger return an error")
+	}
+
+	if v := i.Id(); v != "12345" {
+		t.Fatalf("CreateTrigger did not set resource id. Id was %s", v)
+	}
+
+	if v := i.Get("title"); v != "trigger" {
+		t.Fatalf("CreateTrigger did not set resource title. title was %s", v)
+	}
+}
+
+func TestReadTrigger(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	m := mock.NewClient(ctrl)
+	i := newIdentifiableGetterSetter()
+	i.SetId("12345")
+
+	expected := zendesk.Trigger{
+		Title:  "trigger",
+		Active: true,
+	}
+	m.EXPECT().GetTrigger(gomock.Any(), gomock.Eq(int64(12345))).Return(expected, nil)
+	if err := readTrigger(i, m); err != nil {
+		t.Fatalf("GetTrigger received an error when calling: %v", err)
+	}
+}
+
+func TestUpdateTrigger(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	m := mock.NewClient(ctrl)
+	i := &identifiableMapGetterSetter{
+		id:              "12345",
+		mapGetterSetter: mapGetterSetter{},
+	}
+
+	m.EXPECT().UpdateTrigger(gomock.Any(), gomock.Eq(int64(12345)), gomock.Any()).Return(zendesk.Trigger{}, nil)
+	if err := updateTrigger(i, m); err != nil {
+		t.Fatalf("updateTrigger returned an error %v", err)
+	}
+}
+
+func TestDeleteTrigger(t *testing.T) {
 	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
 	c := mock.NewClient(ctrl)
 	d := newIdentifiableGetterSetter()
 
 	d.SetId("1234")
 
-	c.EXPECT().DeleteTrigger(gomock.Eq(int64(1234))).Return(nil)
-	err := resourceZendeskTriggerDelete(d, c)
+	c.EXPECT().DeleteTrigger(gomock.Any(), gomock.Eq(int64(1234))).Return(nil)
+	err := deleteTrigger(d, c)
 	if err != nil {
 		t.Fatalf("Got error from resource delete: %v", err)
 	}
 }
+
+func testTriggerDestroyed(s *terraform.State) error {
+	client := testAccProvider.Meta().(zendesk.TriggerAPI)
+
+	for k, r := range s.RootModule().Resources {
+		if r.Type != "zendesk_trigger" {
+			continue
+		}
+
+		id, err := atoi64(r.Primary.ID)
+		if err != nil {
+			return err
+		}
+
+		ctx := context.Background()
+		_, err = client.GetTrigger(ctx, id)
+		if err == nil {
+			return fmt.Errorf("did not get error from zendesk when trying to fetch the destroyed trigger. resource name %s", k)
+		}
+
+		zdresp, ok := err.(zendesk.Error)
+		if !ok {
+			return fmt.Errorf("error %v cannot be asserted as a zendesk error", err)
+		}
+
+		if zdresp.Status() != http.StatusNotFound {
+			return fmt.Errorf("did not get a not found error after destroy. error was %v", zdresp)
+		}
+	}
+	return nil
+}
+
+func TestAccTriggerExample(t *testing.T) {
```

Comment:
Im getting a failure on this guy.

```
--- FAIL: TestAccTriggerExample (4.76s)
    testing.go:615: Step 0 error: Check failed: Check 3/4 error: zendesk_trigger.auto-reply-trigger: Attribute 'all' expected to be set
```

Reply:
I was able to get around the white space error by adding a diff suppression func. Here is what i came up with.

```go
"value": {
	Type:     schema.TypeString,
	Required: true,
	DiffSuppressFunc: func(k, old, new string, d *schema.ResourceData) bool {
		if new != "" {
			var newList, oldList []interface{}
			err := json.Unmarshal([]byte(new), &newList)
			if err != nil {
				return false
			}
			err = json.Unmarshal([]byte(d.Get(k).(string)), &oldList)
			if err != nil {
				return false
			}
			return reflect.DeepEqual(newList, oldList)
		}

		return false
	},
},
```

Reply:
Now i am getting this when i run the test.
```
=== RUN   TestAccTriggerExample
--- FAIL: TestAccTriggerExample (29.37s)
    testing.go:615: Step 0 error: errors during apply:
        
        Error: 422: {"error":"RecordInvalid","description":"Record validation errors","details":{"base":[{"description":"The <strong>Email user</strong> cannot be blank"}]}}
        
          on /var/folders/3n/2m9krr0j0gbcbw02v755nz240000gn/T/tf-test273728899/main.tf line 7:
          (source code not available)
        
        
FAIL

Debugger finished with exit code 0
```

Reply:
@tamccall 
Thank you for reporting the acc result.

> zendesk seems to bee messing with the whitespace in the action value which is causing TF to think there is a diff

yeah, I got same result. It could be solved by HCL's `jsonencode()` function like this.

```hcl
  action {
    field = "notification_user"
    value = jsonencode([
      "requester_id",
      "Dear my customer",
      "Hi. This message was configured by terraform-provider-zendesk."
    ])
  }
```

-----

> you should mark the position variable as computed.

Fixed at 5c55b14

-----

> Error: 422: {"error":"RecordInvalid","description":"Record validation errors","details":{"base":[{"description":"The <strong>Email user</strong> cannot be blank"}]}}

I couldn't reproduce this error. It seems to pass all tests now.
Would you try again for `fix-trigger-acc-test` branch?

```
~/workspace/github.com/nukosuke/terraform-provider-zendesk · (fix-trigger-acc-test±)
⟩ go test -v -run "TestAcc.*" ./...
?       github.com/nukosuke/terraform-provider-zendesk  [no test files]
=== RUN   TestAccTicketFieldDataSource
--- PASS: TestAccTicketFieldDataSource (2.52s)
=== RUN   TestAccZendeskAttachment
--- PASS: TestAccZendeskAttachment (3.72s)
=== RUN   TestAccZendeskBrand
--- PASS: TestAccZendeskBrand (3.10s)
=== RUN   TestAccGroupExample
--- PASS: TestAccGroupExample (2.26s)
=== RUN   TestAccTicketFieldExample
--- PASS: TestAccTicketFieldExample (6.36s)
=== RUN   TestAccTicketFormExample
--- PASS: TestAccTicketFormExample (9.15s)
=== RUN   TestAccTriggerExample
--- PASS: TestAccTriggerExample (2.10s)
PASS
ok      github.com/nukosuke/terraform-provider-zendesk/zendesk  29.252s
```

Reply:
👍